### ISSUE TITLE
Recommended way to partially load weights into a model ?

### ISSUE THREAD

**Issue by paganpasta** [patrick-kidger/equinox#136]
Hi,

Is there a recommended way to partially load weights to a new model ? 

At the moment I have an inept solution which uses the fact that `fc` attribute is the only `eqx.nn.Linear` instance in my model

```
is_leaf = lambda x : True if isinstance(x, eqx.nn.Linear) else False
new_model = tree_deserialise_leaves('model.eqx', CNN(num_classes=26, key=key), is_leaf=is_leaf)
```
This approach won't work where there are multiple `Linear` layers. Is there a recommended way to perform such loading? 

A `name` parameter for modules might be useful here.

**Comment by patrick-kidger** [comment]
Yep -- the `filter_spec` argument to `tree_deserialise_leaves` may be a PyTree that is a prefix of the one being loaded, and so you can pass a custom `filter_spec` that does/doesn't load the weights for each part of the model.

Admittedly this does mean you probably want to end up constructing a PyTree with the current `_default_deserialise_filter_spec` on the leaves where you want to load stuff, which is of course a private function. Having ways to more easily extend this default, rather than just overwriting it entirely, would be good.

**Comment by paganpasta** [comment]
Thanks for the suggestion.

Possibly a method which works on a  `PyTree`, `list of leaves` to be ignored and `list of overrides` can be used by the user for deserialization/serialization ?

**Comment by paganpasta** [comment]
@patrick-kidger  having a `strict=True/False` like functionality akin to Pytorch's `load_state_dict` can also be useful in loading partial weights which returns the mismatched leaves along with partially loaded model?

**Comment by patrick-kidger** [comment]
Yeah, that sounds reasonable. Although PyTorch and Equinox serialise in pretty different ways, I'm not sure what a partial match might mean here. That one PyTree is a prefix of the other?

**Comment by paganpasta** [comment]
Yes.

**Comment by patrick-kidger** [comment]
This is surprising tricky. I've not managed to figure out a way to obtain the largest common prefix, using just JAX's public API.

**Comment by paganpasta** [comment]
I see, thanks. 
Happy to work on `strict` and/or extending `_default_deserialise_filter_spec` if you see the advantage of adding them.

**Comment by patrick-kidger** [comment]
Yep, go for it. I think both partial deserialisation and user extensibility are worth handling.

See also the discussion in https://github.com/google/jax/issues/11210 for context.

I'd be happy to contemplate a larger rewrite of these functions if would accomplish these goals. The only thing I'd say is that I think it's important not to have a global registry (as in tjax's single-dispatch approach linked in the link above) as otherwise this opens up the possibility of type piracy.

**Comment by paganpasta** [comment]
Hi @patrick-kidger 

Just wanted to go over some design minutes for extensibility of `_default_...`. I am not familiar with `type piracy` and found resources only discussing it in the context of `Julia`. From what I understood, `f(x)` implemented by our package would be type piracy if neither `f` or `type(x)` are owned by us. So hopefully the proposed approach does not introduce it. 

```python

class DeserializeFilterSpec:
    def __init__(self):
        self.load_jnp = lambda f, x: jnp.load(f)
        self.load_np = lambda f, x: np.load(f)
        self.load_scalar = lambda f, x: np.load(f).item()
        self.load_state = lambda f, x: experimental.set_state(x, jnp.load(f)) or x  #Handling in-place update
        self.load_others = lambda f, x: x

    def make_spec(self):
        def filter_spec(f, x):
            if isinstance(x, jnp.ndarray):
                return self.load_jnp(f, x)
            elif isinstance(x, np.ndarray):
                return self.load_np(f, x)
            elif isinstance(x, (bool, float, complex, int)):  #Can use np.isscalar?
                return self.load_scalar(f, x)
            elif isinstance(x, experimental.StateIndex):
                return self.load_state(f, x)
            else:
                return self.load_others(f, x)
        return filter_spec
```
 This is working with test_serialization.py at the moment for default behaviour. For extension, a user can set appropriate methods. Very roughly like

```python
"""NOT TESTED"""
filter_spec = jax.tree_map(lambda _: DeserializeFilterSpec().make_spec(), like)
custom_spec = DeserializeFilterSpec()
custom_spec.load_jnp = lambda f, x: -x
filter_spec = eqx.tree_at(
        lambda tree: (tree.fc.weight),
        filter_spec,
        replace=(custom_spec.make_spec()),
    )

```
 Alternatively, I can submit a PR with initial solution and we can discuss over there?

**Comment by patrick-kidger** [comment]
I think having a class like this is probably overkill, and would introduce too much user complexity for a relatively niche feature. At least for the problem of user-extensibility, I'm thinking it would probably suffice to just document and make public `_default_{de,}serialise_filter_spec`, so that they can be composed with a la

```python
def filter_spec(f, x):
  if isinstance(x, Foo):
    ...
  else:
    return default_deserialise_filter_spec(f, x)
```

Regarding type piracy: yep, that's pretty much it. If there was a global registry of de/serialisation rules, one may imagine a user registering a de/serialisation rule for dataclasses, because they want to work with them. However this may break some other part of the code (written by a different author, perhaps in a library somewhere) which is relying on dataclasses having different behaviour.

This potential problem is an instance of a general family of errors (most easily found in Julia): type piracy. In this case de/serialisation is a concept owned by Equinox whilst for example dataclasses are a concept owned by the Python standard library. It it only acceptable to register one with the other if you are the owner of one of these concepts.

Incidentally Rust has something very similar: it disallows (with a compile-time error) declaring that a type is of a particular trait, unless you own either the type or the trait.

**Comment by paganpasta** [comment]
Thanks for the detailed response on `type piracy`.

**Comment by paganpasta** [comment]
Hi @patrick-kidger 

Regarding the `strict` functionality to allow loading a prefix of the stored tree. It seems that the current code already does this. 
https://github.com/patrick-kidger/equinox/blob/a89d5b486d13588caffc095f172a2ec39fd68278/equinox/serialisation.py#L185-L187
If `like` is a prefix, then `_assert_same` can only processes this `prefix`. Perhaps, only thing to do here is to reflect this in the documentation of the method which currently states
> `like`: A PyTree of the same structure, and with leaves of the same type, as the PyTree being loaded...

Pushing `strict=False` a bit further one can allow changes in shapes and types thus requiring a lenient or no assert. But, in this case, notifying back the altered leaves would be important to track the changes done to `like`.

**Comment by patrick-kidger** [comment]
Actually, I think at the moment if `like` is a prefix of the saved PyTree then the deserialisation will just go wrong. If the original PyTree has `n` leaves and `like` has `m < n` leaves then at the moment we just try to deserialise the first `m` leaves from the file object `f`. (And this may or may not happen silently, since we use the leaf type in `like` to determine which deserialisation function to use, which may not align with what is actually saved.)

Trying to catch this and raise an error seems like it might be worthwhile, although I'm not completely certain how we might detect this robustly. (We could certainly check that the number of saved leaves matches the number of leaves requested to be loaded. And by serialising strings, perhaps also that the name of their type at save-time matches the name of name of their type and load-time?)

Modulo the above issues, I think I agree that handling `strict` should largely be a matter of adjusting the assert statement.

**Comment by paganpasta** [comment]
So what I meant with `prefix` loading currently works (for same save and load type) was this:
I modified the test from
https://github.com/patrick-kidger/equinox/blob/a89d5b486d13588caffc095f172a2ec39fd68278/tests/test_serialisation.py#L40-L49
to
```python
like = (
        like_jax_array1,
        like_jax_array2,
        like_numpy_array1
)
```
and the leaves are still loaded and the modified assert statement passes.
```python
assert eqx.tree_equal(tree[:3], tree_loaded)
```
The confusion I had was, as per the parameter detail, `like` is meant to be same structure as `tree` which is currently invalid. But, deserialisation still happens. Unless I misunderstood the concept of a `prefix`.

**Comment by patrick-kidger** [comment]
Right -- so this doesn't count as a PyTree prefix, and the fact that that works here is weird (and something we should think about handling better).

Given a PyTree structure like `(*, *, [*])` then a valid prefix would be `(*, *, *)`, i.e. something obtained by replacing a subnode with a leaf. However `(*, *)` would not be a valid prefix: this has not only replaced a subnode with a leaf (as above), it's then mutated the parent node from `(*, *, *)` to `(*, *)`.

As another example in the same vein, then `{"a": *}` isn't a prefix of `{"a": *, "b": *}`.

We can verify this interpretation like so: `jax.tree_util.tree_map(print, (1, 2), (3, 4, 5))` gives `ValueError: Tuple arity mismatch: 3 != 2; tuple: (3, 4, 5).`

So far as I know this isn't actually written down anywhere, and I'm sure the JAX folks would welcome something explicitly specifying this in the PyTree docs. It's certainly an easy enough mistake to make!

**Comment by paganpasta** [comment]
Thanks for the info on the prefix!
I'll try to dig in deeper as to why this case worked for deserialisation.

**Comment by patrick-kidger** [comment]
Something I spotted in parallel alongside this discussion: we're currently using `jax.tree_map` to do de/serialisation. It turns out that this doesn't actually have a guarantee on the order in which it traverses leaves. (In principle it could pick a random order each time.) Fortunately `jax.tree_flatten` does offer a guarantee on the order. As the de/serialisation routines depend on this order being the same between serialisation and deserialisation, we should switch to an `ordered_tree_map` implemented something like:
```python
import jax.tree_util as jtu

def ordered_tree_map(f, x):
  leaves, treedef = jtu.tree_flatten(x)
  leaves = list(map(f, leaves))
  return jtu.tree_unflatten(leaves, treedef)
```
+a bit to handle the additional arguments to `tree_flatten`, which I think is actually mildly complicated. Possibly something like the following:
```python
import jax.tree_util as jtu

def ordered_tree_map(f, x, *others, is_leaf=None):
  y = jtu.tree_map(lambda *a: lambda: f(*a), x, *others, is_leaf=is_leaf)
  leaves, treedef = jtu.tree_flatten(y)
  leaves = [leaf() for leaf in leaves]
  return jtu.tree_unflatten(treedef, leaves)
```
(untested)

**Comment by paganpasta** [comment]
Hi,

As per the implementation of `tree_map`: https://jax.readthedocs.io/en/latest/_modules/jax/_src/tree_util.html#tree_map
It also uses `tree_flatten` on the `tree` and on the `rest` which is later combined with `unflatten`. How does `ordered...` differ from `tree_map`? Is the difference because of two stepped flatten as opposed to single shot using `tree_map` as yours?
```python
  leaves, treedef = tree_flatten(tree, is_leaf)
  all_leaves = [leaves] + [treedef.flatten_up_to(r) for r in rest]
  return treedef.unflatten(f(*xs) for xs in zip(*all_leaves))
```
 Also, could you please point me towards the discussion on guarantee of tree traversal? I'd like to read up more on this. 

I tried plugging in the `ordered..` and it seems to be working fine (existing tests passing for {de}serialisation). Any clues as to how to check that the order is always maintained? I suppose that would require to find a case where the ordering differs between the two `map` approaches. Maybe, the discussion on traversal guarantee will be helpful in this. 

Thanks.

**Comment by patrick-kidger** [comment]
Right -- so the answer is that whilst `tree_map` just-so-happens to be ordered, this is actually an implementation detail. That is, JAX reserves the right to change this behaviour about ordering in the future.

In contrast `tree_flatten` does guarantee that it will return things in order, and JAX actually includes a test that this is the case. This is from here: https://github.com/google/jax/pull/11658 (Which as you can see is very recent. The context is that I raised this question internally with the JAX team, and it was agreed to guarantee that `tree_flatten` will preserve ordering, but not to offer the same guarantee for `tree_map` .)

The current implementation of `tree_map` is a lot neater than my approach though! I suggest we copy-paste that for our `ordered_tree_map`, instead of the one I wrote.

**Comment by paganpasta** [comment]
Hi,
Wanted to double check the behaviour of `StateIndex` within `ordered_tree_map` or current `tree_map`. 
```python
state_1 = eqx.experimental.StateIndex()
state_2 = eqx.experimental.StateIndex()
eqx.experimental.set_state(state_1, jnp.asarray(4))
eqx.experimental.set_state(state_2, jnp.asarray(2))
z = jtu.tree_map(lambda *xs: tuple(xs), state_1, state_2)
``` 
currently fails with 
```
ValueError: Mismatch custom node data: 
(('_state', 'inference'), ('_obj', '_version'), (<equinox.experimental.stateful._IndexObj object at 0x7f0693a31c70>, <equinox.experimental.stateful._FixedInt object at 0x7f06939fbd00>)) 
!= 
(('_state', 'inference'), ('_obj', '_version'), (<equinox.experimental.stateful._IndexObj object at 0x7f0693a31ee0>, <equinox.experimental.stateful._FixedInt object at 0x7f06939fbc70>)); 
value: StateIndex(inference=False).
```
The presence of static values of `_obj`, `_version` seems to be triggering this error. Looks like an unexpected behaviour as other Objects work fine for leaves.

**Comment by patrick-kidger** [comment]
This looks expected. Each `StateIndex` has different static values to every other state index. And thus a different PyTree structure.

In practice the de/serialisation routines do understand `StateIndex` and are able to handle it, just like they're capable of understanding other custom types.

**Comment by paganpasta** [comment]
Thanks for confirming this.

**Comment by paganpasta** [comment]
Unsurprisingly, `ordered_tree_map` fits in well. I wasn't sure if this should be made public or not. At the moment, it is a public method. 
As per previous discussions [here](https://github.com/patrick-kidger/equinox/issues/136#issuecomment-1196469019)
>  Actually, I think at the moment if like is a prefix of the saved PyTree then the deserialisation will just go wrong. If the original PyTree has n leaves and like has m < n leaves then at the moment we just try to deserialise the first m leaves from the file object f. (And this may or may not happen silently, since we use the leaf type in like to determine which deserialisation function to use, which may not align with what is actually saved.)

> Trying to catch this and raise an error seems like it might be worthwhile, although I'm not completely certain how we might detect this robustly. (We could certainly check that the number of saved leaves matches the number of leaves requested to be loaded. And by serialising strings, perhaps also that the name of their type at save-time matches the name of name of their type and load-time?)

> Modulo the above issue, I think I agree that handling strict should largely be a matter of adjusting the assert statement.

We do need some additional information to be stored during serialisation which can notify us about a potential mismatch. Perhaps, storing part of `tree_flatten` information in some way which can easily be parsed? It does contain leaf information and tree definition as well.

**Comment by patrick-kidger** [comment]
I'd leave it private for now; it's not important enough to be worth documenting IMO.

Hmm yeah, so this looks like a tricky issue. Really the whole point of the `like` argument is to allow us *not* to have to serialise anything else, since doing so (pickling etc.) is so error-prone.

Perhaps, instead of trying to implement this feature, the recommend approach here should be to instead deserialise the bit you care about, and then do model surgery:
```python
model = CNN()
linear = eqx.tree_deserialise_leaves("model.eqx", model.final_linear)
model = eqx.tree_at(lambda m: m.final_linear, model, linear)
```

**Comment by paganpasta** [comment]
Sorry I did not understand it completely. 
So `model.eqx` is the serialised complete model.
`model.final_linear` will be saved towards the end of the serialised file assuming it happens to be the last (leaf) defined parameter.
`tree_deserialise...` iterates over leaves of `like` and streams from the saved file. We need to sort of 'skip' to the saved location in the `model.eqx` for the `final_linear`?

**Comment by patrick-kidger** [comment]
Sorry, I reversed your original example. I think your original comment was about deserialising everything except the final linear layer? (My code was about deserialising just the final linear layer.)

So in this case you could do:
```python
model_original = CNN()
model = eqx.tree_deserialise_leaves("model.eqx", model_original)
model = eqx.tree_at(lambda m: m.final_linear, model, model_original)
```

My point, now that we've debated this back and forth for a bit, is that I don't see an obvious way to implement this at the level of `tree_deserialise_leaves`. So, perhaps the recommended approach here really should be to do something like in your original comment, or in my snippet above.

**Comment by paganpasta** [comment]
Ahh, I see. I think your proposed approach is more elegant. 
Sorry to stretch this discussion so much. 
Hopefully, this is the last bit in this thread:
 
I don't think this case can be identified at the moment https://github.com/patrick-kidger/equinox/issues/136#issuecomment-1196522572 by the current `deserialize...` implementation. Probably a documentation update to reflect it is better?

**Comment by patrick-kidger** [comment]
Yep, I could see some documentation being a good idea. (Including a version of my above example?)

**Comment by vvvm23** [comment]
Sorry to necro an old thread (is necro a thing on Github :thinking:) but this seemed related and still opened. If you prefer I can open another issue.

I've been working on a latent video diffusion project, which involves training a LDM on image frames as a first-stage. Then, I insert temporal layers into each block which I implement by subclassing the original blocks. The rough structure can be outlined like so:

```python
import equinox as eqx
import equinox.nn as nn
import jax

from typing import List

class Layer(eqx.Module):
    fc1: nn.Linear

    def __init__(self, key):
        self.fc1 = nn.Linear(2, 2, key=key)

    def __call__(self, x):
        return self.fc1(x)

class ExtendedLayer(Layer):
    fc2: nn.Linear

    def __init__(self, key):
        key, subkey = jax.random.split(key)
        super().__init__(subkey)

        self.fc2 = nn.Linear(2, 2, key=key)

    def __call__(self, x):
        x = super().__call__(x)
        return self.fc2(x)

class Model(eqx.Module):
    layers: List[Layer]

    def __init__(self, key, extend: bool = False):
        self.layers = []
        
        init_layer = ExtendedLayer if extend else Layer 
        for _ in range(2):
            key, subkey = jax.random.split(key)
            self.layers.append(init_layer(subkey))

    def __call__(self, x):
        for l in self.layers:
            x = l(x)

        return x
```

I have trained a model with `extend=False`, and want to load keys from it into a new model with `extend=True`, populating matching keys and leaving unmatching randomly initialised.  However, clearly the pytree structure is different, so it is hard to use `deserialise_tree_leaves` in this case:
```python
key = jax.random.PRNGKey(0xABC)

# in reality, this is loaded from `eqx.deserialise_leaves`, but just init here for simplicity.
trained_model = Model(key=key)

key, model_key = jax.random.split(key)
extended_model = Model(key=model_key, extend=True)

print(trained_model)
print(extended_model)

===
Out:
Model(
  layers=[
    Layer(
      fc1=Linear(
        weight=f32[2,2],
        bias=f32[2],
        in_features=2,
        out_features=2,
        use_bias=True
      )
    ),
    Layer(
      fc1=Linear(
        weight=f32[2,2],
        bias=f32[2],
        in_features=2,
        out_features=2,
        use_bias=True
      )
    )
  ]
)

Model(
  layers=[
    ExtendedLayer(
      fc1=Linear(
        weight=f32[2,2],
        bias=f32[2],
        in_features=2,
        out_features=2,
        use_bias=True
      ),
      fc2=Linear(
        weight=f32[2,2],
        bias=f32[2],
        in_features=2,
        out_features=2,
        use_bias=True
      )
    ),
    ExtendedLayer(
      fc1=Linear(
        weight=f32[2,2],
        bias=f32[2],
        in_features=2,
        out_features=2,
        use_bias=True
      ),
      fc2=Linear(
        weight=f32[2,2],
        bias=f32[2],
        in_features=2,
        out_features=2,
        use_bias=True
      )
    )
  ]
)
```

I investigated the approach you two outlined with `tree_at` but I don't think it works here, as the Pytrees have different structures. Do you have any advice for the partial loading of weights for models with different, but highly similar, structure?

**Comment by patrick-kidger** [comment]
Can you:
(a) construct a `like` that is of the correct structure (by creating a dummy `Model(extend=False)`), and use that to deserialise;
(b) construct a model that is of the desired final structure (`Model(extend=True)`);
(c) Use `eqx.tree_at` to move the weights from the first to the second?

Side note, I'd recommend against subclassing a concrete class. (Can you compose them instead?)
It's not an advertised part of the public docs yet, but I've got a draft of the [abstract/final design pattern here](https://docs.kidger.site/equinox/pattern/), that you might find interesting.

**Comment by vvvm23** [comment]
(a) and (b) is kinda what I have already, being lines `trained_model = ...` and `extended_model = ...` respectively - just imagine (a) is loaded with the trained weights.

The tricky part is (c). It is possible to construct it for this specific example like so:

```python
def where_fn(tree):
    return jtu.tree_leaves(
        jtu.tree_map(
            lambda v: v.fc1 if hasattr(v, 'fc2') else v, 
            tree, 
            is_leaf=lambda v: isinstance(v, Layer))
        )

extended_model = eqx.tree_at(where_fn, extended_model, jtu.tree_leaves(trained_model))
assert jax.numpy.allclose(extended_model.layers[0].fc1.weight, trained_model.layers[0].fc1.weight)
assert jax.numpy.allclose(extended_model.layers[1].fc1.weight, trained_model.layers[1].fc1.weight)
```

Which works fine, but I guess I wasn't precise enough in my question.

What I intended is *is there a way to do this in the general case?* As in, given two pytrees / Equinox models, populate the target leafs if and only if a target leaf of the same path exists in the other pytree.

The above is a generalisation of finetuning via adding additional layers, which includes adding a layer to the end of the model (like originally in this issue), inserting extra layers like I am doing, or something like LoRA adapters.

**Comment by patrick-kidger** [comment]
In the general case, you could probably do this with a bit of work using `jax.tree_util.tree_flatten_with_path(model_one)`, considering each path, and following it for `model_two` until you either get a leaf or hit an error.

If the case that you control the model construction as well, then as an alternative you could probably arrange to make them `tree_map`-compatible with each other. To be precise: have `tree_structure(original_model, is_leaf=lambda x: isinstance(x, (...some collection of layers...))` be a prefix of `tree_structure(extended_model, is_leaf=lambda x: isinstance(x, (...some collection of layers...))`.
